number/src/baby_bear.rs

```diff
@@ -0,0 +1,467 @@
+use p3_baby_bear::BabyBear;
+use schemars::{
+    schema::{Schema, SchemaObject},
+    JsonSchema,
+};
+use serde::{Deserialize, Serialize};
+
+use num_traits::{ConstOne, ConstZero};
+use std::ops::{Add, AddAssign, Div, Mul, MulAssign, Neg, Not, Sub, SubAssign};
+use std::str::FromStr;
+use std::{collections::BTreeSet, fmt::LowerHex};
+
+use crate::{BigUint, FieldElement, KnownField, LargeInt};
+use ark_ff::{One, Zero};
+
+use core::fmt::{self, Debug, Formatter};
+use core::hash::Hash;
+
+use p3_field::{AbstractField, Field, PrimeField32};
+
+#[derive(
+    Debug,
+    Copy,
+    Clone,
+    Default,
+    Eq,
+    Hash,
+    PartialEq,
+    Ord,
+    PartialOrd,
+    Serialize,
+    Deserialize,
+    derive_more::Display,
+)]
+pub struct BabyBearField(BabyBear);
+
+const P: u32 = 0x78000001;
+const MONTY_BITS: u32 = 32;
+const MONTY_ZERO: u32 = to_monty(0);
+const MONTY_ONE: u32 = to_monty(1);
+
+const fn to_monty(x: u32) -> u32 {
+    (((x as u64) << MONTY_BITS) % P as u64) as u32
+}
+
+impl BabyBearField {
+    const ORDER: u32 = P;
+
+    #[inline(always)]
+    fn from_canonical_u32(n: u32) -> Self {
+        Self(BabyBear::from_canonical_u32(n))
+    }
+
+    #[inline]
+    fn to_canonical_u32(self) -> u32 {
+        self.0.as_canonical_u32()
+    }
+}
+
+impl FieldElement for BabyBearField {
+    type Integer = BBLargeInt;
+
+    const BITS: u32 = 31;
+
+    fn to_degree(&self) -> crate::DegreeType {
+        self.to_canonical_u32() as u64
+    }
+
+    fn to_integer(&self) -> Self::Integer {
+        self.to_canonical_u32().into()
+    }
+
+    #[inline]
+    fn modulus() -> Self::Integer {
+        Self::ORDER.into()
+    }
+
+    fn pow(self, exp: Self::Integer) -> Self {
+        Self(BabyBear::exp_u64_generic(
+            self.0,
+            exp.try_into_u64().unwrap(),
+        ))
+    }
+
+    fn to_bytes_le(&self) -> Vec<u8> {
+        self.to_canonical_u32().to_le_bytes().to_vec()
+    }
+
+    fn from_bytes_le(bytes: &[u8]) -> Self {
+        let u = u32::from_le_bytes(bytes.try_into().unwrap());
+        Self::from_canonical_u32(u)
+    }
+
+    fn from_str_radix(s: &str, radix: u32) -> Result<Self, String> {
+        u32::from_str_radix(s, radix)
+            .map(Self::from_canonical_u32)
+            .map_err(|e| e.to_string())
+    }
+
+    fn checked_from(value: ibig::UBig) -> Option<Self> {
+        if value < Self::modulus().to_arbitrary_integer() {
+            Some(u32::try_from(value).unwrap().into())
+        } else {
+            None
+        }
+    }
+
+    fn is_in_lower_half(&self) -> bool {
+        self.to_canonical_u32() <= (Self::ORDER - 1) / 2
+    }
+
+    fn known_field() -> Option<crate::KnownField> {
+        Some(KnownField::BabyBearField)
+    }
+
+    fn try_into_i32(&self) -> Option<i32> {
+        Some(self.to_canonical_u32() as i32)
```

Comment:
I don't think this is the right implementation. It should return `-1` for `bb(0) - bb(1)`.

Reply:
btw these files were git's merge mess, they were already merged in https://github.com/powdr-labs/powdr/pull/1696

I'll fix it in another PR